<a href="https://colab.research.google.com/github/somasundar0708/Pizza-insights-project/blob/main/phyton.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
6# bro code : 5 video
friends = 0
friend = friends + 1 #option 1
friends += 1 # option 2 but both are same action will do """This is called :Aurgmented assiement
print(friend)# we can try with "-,*,/,** = power,// = floor division,% = remainder"

1


In [ ]:
#built in function
x = 3.94
d = 8
f = 12
#result = round(x)# Round funtion : it will print nerest value in float
# result = abs(d - f) # Absolute funtion : it will print only positive value
#result = max(d,f) # Max funtion : it will print max value
result = min(d,f,x) # Min funtion : it will print min value
#result =pow(d,f) # power function
print(result)

3.94


In [ ]:
#math module
import math
print(math.pi)
result = print(math.sqrt(9))
print(math.ceil(3445.43543))
print(math.floor(23987.324))

#Exercise find the circumference
radius = int(input("Enter the radius : "))
circumference = 2 * math.pi * radius
print(f"circumference is {round(circumference)}")

# Exercise area circle
radius = int(input("Enter the radius : "))
area = math.pi * radius ** 2
print(f"the are is {area}")

# Hypotenuse of right triangle
base = int(input("Enter the base : "))
height = int(input("Enter the height : "))



3.141592653589793
3.0
3446
23987
Enter the radius : 4e2


ValueError: invalid literal for int() with base 10: '4e2'

In [ ]:
# if,else,elif condition
age = int(input("Enter your age : "))
if age >= 45:     # 1)condition must be true its acts boolean datatype 2)indention must necessary 3)elif used in multiple option
  print("too old")
elif age >= 18:
  print("welcome")
else :
  print("too young")

In [ ]:
sun = int(input("enter the number :" ))
dar = int(input("enter the number :"))
hi = input("enter operator :" )
if hi == "+":
  print(sun + dar)
elif hi == "-":
  print(sun - dar)
elif hi == "*":
  print(sun * dar)
elif hi == "//":
  print(sun // dar)
else:
  print("invalid")

In [ ]:
sun = 1
while sun <= 10:
  print(sun)
  sun += 1

In [ ]:
sun = 2
while sun <= 20:
  if sun % 2 == 0:
    print(sun)
  sun += 1

In [ ]:
sun = input("enter your name :")
while sun == "":
   sun = input("enter your name buddy : ")
print(f"hello  {sun}")

In [ ]:
# Logical operator = usedd on conditional statement.
#and = both condition must true
# or = any one condition must be true.
# not = if condition true it print flase like reverse
temp = 25
if temp > o and temp < 30:
  print("the temp is good")
else :
  print("bad temp")

In [ ]:
# Conditional expersion
num = 5
a=3
k=5
age = 13
user = True
#print("positive" if num > 0 else "negetive")
#result = "nfv" if num < 0 else "rdf"
#print(result)
#max = a if a > k else k
#status = "hi" if age >= 18 else "byr"
hi = "hi" if user else "bye"
print(f"run {hi}")


In [ ]:
# String methods only ibn string methods
name = input("enter your name : ")
number = input("enter phonr number : ")
#result = len(name) length will start count 1
#result = name.find("d")#print the index of the letter. index will start count with 0
#result = name.rfind("a")
#result= name.capitalize()# starting letter capital
#result = name.upper()#Can we put something inside the brackets? No.The upper() method does not accept any arguments.
# we can use arugments in upper,lower,capitalize,isdight
#| Method        | Checks                   |
#| ------------- | ------------------------ |
#| `isdigit()`   | Only digits (0–9)        |
#| `isnumeric()` | Digits + numeric symbols |
#| `isalpha()`   | Only letters             |
#| `isalnum()`   | Letters + digits         |
#result = number.count("9")
result = number.replace("/","--")
print(result)
print(help(int))


In [ ]:
user = input("Enter Your Name : ")
if (len(user)<=12 and user != "") and user != "0-9":
  print(f"welcome {user}")
else :
  print("invaild username")

Enter Your Name : 0 soma sundarfr
invaild username


In [ ]:
# String index in python
# Indexing =  accessing elements of a sequence using [],,,[start : end : step]
credit_number = "1234-5678-9012-3456"
#print(credit_number[1])
#print(credit_number[0:2])
#print(credit_number[:8])
#print(credit_number[5:])
#print(credit_number[-1])
#print(credit_number[::2])
# i want last four digit number
#last4 = credit_number[15:11]
#print(last4)
hi = credit_number[::-1] # Reverse order
print (hi)


6543-2109-8765-4321


In [ ]:
# Email slicer prog5ram
email = input("enter you mail : ")
index = email.index("@")
sun = email[:index]
dar = email[index:]
print(f"your mai is {dar} and domain is {sun}")


enter you mail : yugfvygf2@jvgjggy
your mai is @jvgjggy and domain is yugfvygf2


In [ ]:

# format specifiers = {value:flags} format a value based on what flags are inserted#
# .(number)f = round to that many decimal places (fixed point)|
# : (number) = allocate that many spaces
# :03 = allocate and zero pad that many spaces
# :< = left justify
# :> = right justify
# :^ = center align
# :+ = use a plus sign to indicate positive value
# := = place sign to leftmost position
# : = insert a space before positive numbers
# :, = comma separator
net = 2323.32254345
net1 = 43534.3465
net3 = 43543.66767
#print(f"{net:.3f}")
#print(f"{net1:.2f}")
#print(f"{net3:.1f}")
print(f"{net:.10}")
print(f"{net1:.8}")
print(f"{net3:.6}")
print(f"{net:.060}")
print(f"{net1:.0550}")
print(f"{net3:.070}")
print(f"{net:<10}")
print(f"{net1:<10}")
print(f"{net3:<10}")


2323.322543
43534.346
43543.7
2323.322543450000011944212019443511962890625
43534.346499999999650754034519195556640625
43543.6676700000025448389351367950439453125
2323.32254345
43534.3465
43543.66767


In [ ]:
# while loop = execute some code while some condition remains true.
#name = input("enter your name :")
#while name == "":
 # print("you did not your name :")
  #name = input("enter your name buddy :")
#print(name)
oder_plz = input()



enter your age :


ValueError: invalid literal for int() with base 10: ''

In [ ]:
while True:
    age = input("Enter your age: ")

    # Case 1: No input
    if age == "":
        print("No input given. Please enter your age.")
        continue

    # Case 2: Non-numeric input
    if not age.isdigit():
        print("Invalid input. Please enter numbers only.")
        continue

    # Convert to integer after validation
    age = int(age)

    # Age conditions
    if age >= 18:
        print("Welcome")
        break
    else:
        print(f"You are not allowed with : {age}")
        break


Enter your age: 
No input given. Please enter your age.
Enter your age: 
No input given. Please enter your age.
Enter your age: @
Invalid input. Please enter numbers only.
Enter your age: 12
You are not allowed with : 12


In [ ]:
while True:
    age = input("Enter your age: ")

    # Case 1: No input
    if age == "":
        print("No input given. Please enter your age.")
        continue

    # Case 2: Non-numeric input
    if not age.isdigit():
        print("Invalid input. Please enter numbers only.")
        continue

    # Convert to integer after validation
    age = int(age)

    # Age conditions
    if age >= 18:
        print("Welcome")
        break
    else:
        print(f"You are not allowed with : {age}")
        break


In [ ]:
# For loop : Execute a block of code a fixed number of times
#            You can iterate over a arange,string,sequence.etc
# iterate : means its kind of loop in sequence method.
#for x in range(1,11):
#  print(x)
# Reversed function
#for x in reversed(range(1,11)):
#  print(x)
# Step method
#for x in range(1,11,-1):
#   print(x)
# Continue : Skip the specific element which we give to if condition range.it mostly used in filtering purpose.
for x in range(1,9):
  if x != 5:
    continue
  print(x)
  i = 0
while i < 5:
    i += 1
    if i == 3:
        continue
    print(i)


5
1
2
4
5


In [ ]:
i = 0
while i < 5:
    i += 1
    if i == 3:
        continue
    print(i)


1
2
4
5


In [ ]:
# Break :break is used to stop a loop when further execution is not needed.
#| Feature                 | `continue` | `break`     |
#| ----------------------- | ---------- | ----------- |
#| Skips current iteration | Yes        | No          |
#| Stops loop completely   | No         | Yes         |
#| Loop continues          | Yes        | No          |
#| Common use              | Filtering  | Termination |
for x in range (1,999):
  if x == 678:
    break
  else:
    print(x)


1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277


Below is a **clear, structured, and exam-oriented explanation** of the **`in` and `is` operators in Python**, including **why, where, and how they are used**, with examples and common mistakes.

---

## 1. `in` Operator (Membership Operator)

### What is `in`?

`in` is a **membership operator**.

* Checks **whether a value exists inside a sequence**
* Returns **True or False**
* Works with: `list`, `tuple`, `set`, `string`, `dictionary`

```python
value in sequence
```

---

### Why do we use `in`?

* To **check presence of an element**
* To **validate user input**
* To **search data**
* To **avoid loops for simple checks**

---

### Examples

#### 1. List

```python
fruits = ["apple", "banana", "mango"]

print("apple" in fruits)   # True
print("orange" in fruits)  # False
```

---

#### 2. String

```python
text = "python programming"

print("python" in text)   # True
print("java" in text)     # False
```

---

#### 3. Dictionary (checks keys only)

```python
data = {"id": 101, "name": "Sundar"}

print("id" in data)       # True
print("Sundar" in data)   # False
```

---

### Real-time use case

```python
while True:
    value = input("Enter q to quit: ")
    if value in ["q", "Q"]:
        break
```

---

## 2. `is` Operator (Identity Operator)

### What is `is`?

`is` is an **identity operator**.

* Checks **whether two variables refer to the same object in memory**
* Returns **True or False**
* Compares **memory location**, not value

```python
a is b
```

---

### Why do we use `is`?

* To check **object identity**
* To compare with **None**
* For **memory-sensitive checks**

---

### Examples

#### 1. Correct use with `None`

```python
x = None

if x is None:
    print("x is None")
```

---

#### 2. `is` vs `==`

```python
a = [1, 2, 3]
b = [1, 2, 3]

print(a == b)  # True  (same values)
print(a is b)  # False (different objects)
```

---

#### 3. Small integers (special case)

```python
a = 10
b = 10

print(a is b)  # True (due to integer caching)
```

⚠ Do not rely on this behavior in logic.

---

## 3. Key Differences (Very Important)

| Feature        | `in`                 | `is`                  |
| -------------- | -------------------- | --------------------- |
| Operator type  | Membership           | Identity              |
| Checks         | Presence             | Same object           |
| Compares       | Value in sequence    | Memory location       |
| Used with      | Collections, strings | Objects, None         |
| Common mistake | —                    | Using instead of `==` |

---

## 4. Common mistakes (Exam-important)

### ❌ Using `is` instead of `==`

```python
a = [1, 2]
b = [1, 2]

if a is b:    # WRONG
    print("Equal")
```

### ✔ Correct

```python
if a == b:
    print("Equal")
```

---

### ❌ Using `in` for identity

```python
if x in None:   # ERROR
    pass
```

---

## 5. One-line exam definitions

* **`in`**: Checks whether a value exists in a sequence.
* **`is`**: Checks whether two variables refer to the same object in memory.

---

## 6. Simple rule to remember

> **Use `in` to check presence, `==` to check value, and `is` to check identity.


nested loop : a loop within another loop
               outer loop:
                 inner loop:
  Below is a **clear, structured, and exam-oriented explanation** of **nested loops in Python**, from **basic to intermediate level**, with examples, use cases, and common mistakes.

---

## 1. What is a Nested Loop?

A **nested loop** is a loop **inside another loop**.

* The **outer loop** runs first
* For **each iteration of the outer loop**, the **inner loop runs completely**

```python
for outer in range(...):
    for inner in range(...):
        # inner loop body
```

---

## 2. Why do we use Nested Loops?

We use nested loops when:

* Working with **2D data** (matrices, tables)
* Generating **patterns**
* Comparing **each element with every other element**
* Handling **rows and columns**
* Solving **DSA problems** (sorting, searching, grids)

In short:

> **Nested loops are used when one iteration depends on another iteration.**

---

## 3. How Nested Loops Work (Execution Flow)

Example:

```python
for i in range(1, 4):
    for j in range(1, 3):
        print(i, j)
```

### Execution:

* `i = 1` → `j = 1, 2`
* `i = 2` → `j = 1, 2`
* `i = 3` → `j = 1, 2`

### Output:

```
1 1
1 2
2 1
2 2
3 1
3 2
```

---

## 4. Nested `for` Loop Example (Pattern)

```python
for i in range(3):
    for j in range(3):
        print("*", end=" ")
    print()
```

Output:

```
* * *
* * *
* * *
```

---

## 5. Nested Loop with `while`

```python
i = 1
while i <= 3:
    j = 1
    while j <= 2:
        print(i, j)
        j += 1
    i += 1
```

Key point:

* Inner loop variable **must be reset** every time

---

## 6. Real-Time Examples

### 1. Multiplication Table

```python
for i in range(1, 4):
    for j in range(1, 11):
        print(i * j, end=" ")
    print()
```

---

### 2. Matrix Traversal

```python
matrix = [[1, 2], [3, 4]]

for row in matrix:
    for value in row:
        print(value)
```

---

### 3. Pair Comparison

```python
numbers = [1, 2, 3]

for i in numbers:
    for j in numbers:
        print(i, j)
```

---

## 7. Nested Loop with `break` and `continue`

### `break` (breaks inner loop only)

```python
for i in range(3):
    for j in range(3):
        if j == 1:
            break
        print(i, j)
```

---

### `continue` (skips inner iteration)

```python
for i in range(3):
    for j in range(3):
        if j == 1:
            continue
        print(i, j)
```

---

## 8. Time Complexity (DSA-important)

| Loops              | Complexity |
| ------------------ | ---------- |
| Single loop        | O(n)       |
| Two nested loops   | O(n²)      |
| Three nested loops | O(n³)      |

Important for **exams and interviews**.

---

## 9. Common Mistakes

1. Forgetting to reset inner loop variable (`while`)
2. Using `break` thinking it stops all loops
3. Infinite loops in nested `while`
4. Wrong indentation

---

## 10. One-line Exam Definition

> **A nested loop is a loop inside another loop, where the inner loop executes completely for each iteration of the outer loop.**

In [ ]:
for x in range(2):
  for j in range(11,20):
     print(j,end="-")
  print()

rows = int(input("enter how many rows: "))
columns = int(input("enter colomus : "))
synbol = input("enter symbol")
for x in range(rows):
  for y in range(columns):
    print(synbol,end= " ")
  print()


11-12-13-14-15-16-17-18-19-
11-12-13-14-15-16-17-18-19-
enter how many rows: 4
enter colomus : 10
enter symbol]
] ] ] ] ] ] ] ] ] ] 
] ] ] ] ] ] ] ] ] ] 
] ] ] ] ] ] ] ] ] ] 
] ] ] ] ] ] ] ] ] ] 


In [ ]:
my_time = int(input("Enter the time in seconds: "))

for x in range(my_time, 0, -1):
   seconds = x % 60
   minutes = int(x / 60) % 60
   hours = int(x / 3600)
   print(f"{hours:02}:{minutes:02}:{seconds:02}")

print("TIME'S UP!")

Enter the time in seconds: 7
00:00:07
00:00:06
00:00:05
00:00:04
00:00:03
00:00:02
00:00:01
TIME'S UP!


Below is a **clear, structured, and exam-oriented explanation** of **Lists in Python**, from **basic to intermediate level**, with syntax, examples, real-time use cases, and common mistakes.

---

## 1. What is a List in Python?

A **list** is a **mutable**, **ordered** collection used to store **multiple values** in a single variable.

```python
my_list = [1, 2, 3, 4]
```

---

## 2. Why do we use Lists?

We use lists when:

* We need to store **multiple values together**
* Data needs to be **changed (add/remove/update)**
* Order of elements **matters**
* Same variable should hold **dynamic data**

In short:

> **Lists are used to store and manage multiple related items efficiently.**

---

## 3. Key Properties of Lists

| Property          | Description                          |
| ----------------- | ------------------------------------ |
| Ordered           | Elements have index positions        |
| Mutable           | Values can be changed                |
| Allows duplicates | Same value can appear multiple times |
| Heterogeneous     | Can store different data types       |

Example:

```python
data = [10, "apple", 3.5, True]
```

---

## 4. Creating Lists

```python
numbers = [1, 2, 3]
names = ["Sundar", "Ravi", "Anu"]
empty = []
```

---

## 5. Accessing List Elements (Indexing)

```python
names = ["Sundar", "Ravi", "Anu"]

print(names[0])   # Sundar
print(names[-1])  # Anu
```

---

## 6. Modifying List Elements

```python
numbers = [1, 2, 3]
numbers[1] = 20
print(numbers)
```

Output:

```
[1, 20, 3]
```

---

## 7. Common List Methods (Very Important)

| Method      | Use                   |
| ----------- | --------------------- |
| `append()`  | Add element at end    |
| `insert()`  | Add at specific index |
| `remove()`  | Remove by value       |
| `pop()`     | Remove by index       |
| `clear()`   | Remove all elements   |
| `sort()`    | Sort list             |
| `reverse()` | Reverse list          |
| `count()`   | Count occurrences     |
| `index()`   | Find index            |

### Example

```python
nums = [3, 1, 2]
nums.append(5)
nums.sort()
print(nums)
```

---

## 8. Looping Through a List

### Using `for` loop

```python
fruits = ["apple", "banana", "mango"]

for fruit in fruits:
    print(fruit)
```

---

### Using `while` loop

```python
i = 0
while i < len(fruits):
    print(fruits[i])
    i += 1
```

---

## 9. Real-Time Use Cases

### 1. Student marks

```python
marks = [78, 85, 90, 66]
average = sum(marks) / len(marks)
```

---

### 2. Shopping cart

```python
cart = []
cart.append("Milk")
cart.append("Bread")
```

---

### 3. Filtering data

```python
numbers = [1, 2, 3, 4, 5]

for n in numbers:
    if n % 2 == 0:
        print(n)
```

---

## 10. List vs Tuple (Exam-Important)

| Feature     | List         | Tuple      |
| ----------- | ------------ | ---------- |
| Mutable     | Yes          | No         |
| Syntax      | `[]`         | `()`       |
| Performance | Slower       | Faster     |
| Use case    | Dynamic data | Fixed data |

---

## 11. Common Mistakes

1. Index out of range ❌
2. Modifying list while iterating ❌
3. Forgetting lists are mutable ❌
4. Using `=` instead of `.copy()` ❌

```python
a = [1, 2]
b = a
b.append(3)
# a is also changed
```

---

## 12. One-line Exam Definition

> **A list is an ordered, mutable collection used to store multiple elements in Python.**

---

## 13. Must-Remember Points for Exams

* Lists are **mutable**
* Indexing starts at **0**
* Supports **slicing**
* Supports **nested lists**

most important

| Method         | Meaning            |
| -------------- | ------------------ |
| `append(x)`    | Add element at end |
| `insert(i, x)` | Add at index       |
| `remove(x)`    | Remove value       |
| `pop(i)`       | Remove by index    |
| `clear()`      | Remove all         |
| `sort()`       | Sort list          |
| `reverse()`    | Reverse order      |
| `count(x)`     | Count value        |
| `index(x)`     | Find index         |


In [ ]:
fruits = ["apple","orange","banana","coconut","pineapple","dragan","kiwi"]
#print(fruits)
#print(fruits[:2:-2])
#for x in fruits:
#  print(x)
#dir(fruits)# its shows Methonds of list
#help(fruits)
len(fruits)
"Apple" in fruits
fruits[3] = "avacado"
print(fruits)
fruits.append("bujji")
print(fruits)
fruits.remove("kiwi")
print(fruits)
fruits.insert(2,"yoyoyoyo")
print(fruits)
fruits.pop(0)

['apple', 'orange', 'banana', 'avacado', 'pineapple', 'dragan', 'kiwi']
['apple', 'orange', 'banana', 'avacado', 'pineapple', 'dragan', 'kiwi', 'bujji']
['apple', 'orange', 'banana', 'avacado', 'pineapple', 'dragan', 'bujji']
['apple', 'orange', 'yoyoyoyo', 'banana', 'avacado', 'pineapple', 'dragan', 'bujji']


'apple'

Below is a **clear, structured, and exam-oriented explanation** of **Sets in Python**, from **basic to intermediate level**, with syntax, properties, methods, examples, and common mistakes.

---

## 1. What is a Set in Python?

A **set** is an **unordered**, **mutable** collection that stores **unique values only**.

```python
my_set = {1, 2, 3}
```

---

## 2. Why do we use Sets?

We use sets when:

* We want to **remove duplicate values**
* Order of elements **does not matter**
* We need **fast membership checking**
* Performing **mathematical set operations** (union, intersection)

In short:

> **Sets are used to store unique elements and perform set operations efficiently.**

---

## 3. Key Properties of Sets (Very Important)

| Property    | Description               |
| ----------- | ------------------------- |
| Unordered   | No index positions        |
| Unique      | No duplicate values       |
| Mutable     | Can add/remove elements   |
| No indexing | Cannot access by index    |
| Fast lookup | Efficient membership test |

Example:

```python
s = {1, 2, 2, 3}
print(s)
```

Output:

```
{1, 2, 3}
```

---

## 4. Creating Sets

```python
numbers = {1, 2, 3}
names = {"Sundar", "Ravi"}
empty_set = set()      # correct
```

⚠️ `{}` creates a **dictionary**, not a set.

---

## 5. Accessing Elements (Important)

* Sets **do not support indexing**
* Use loops or membership operators

```python
s = {10, 20, 30}

for x in s:
    print(x)

print(20 in s)   # True
```

---

## 6. Common Set Methods

### Adding elements

```python
s.add(4)
```

### Removing elements

```python
s.remove(2)     # error if not present
s.discard(2)    # no error
s.pop()         # removes random element
```

### Clear set

```python
s.clear()
```

---

## 7. Set Operations (Exam-Important)

```python
a = {1, 2, 3}
b = {3, 4, 5}
```

| Operation            | Method  | Result      |               |
| -------------------- | ------- | ----------- | ------------- |
| Union                | `a      | b`          | `{1,2,3,4,5}` |
| Intersection         | `a & b` | `{3}`       |               |
| Difference           | `a - b` | `{1,2}`     |               |
| Symmetric Difference | `a ^ b` | `{1,2,4,5}` |               |

---

## 8. Real-Time Use Cases

### 1. Remove duplicates from list

```python
nums = [1, 2, 2, 3, 3]
unique = set(nums)
```

---

### 2. Common elements

```python
students_A = {"A", "B", "C"}
students_B = {"B", "C", "D"}

print(students_A & students_B)
```

---

### 3. Fast lookup

```python
allowed_users = {"admin", "user"}

if "admin" in allowed_users:
    print("Access granted")
```

---

## 9. Set vs List (Important)

| Feature     | Set           | List   |
| ----------- | ------------- | ------ |
| Order       | No            | Yes    |
| Duplicates  | No            | Yes    |
| Indexing    | No            | Yes    |
| Performance | Faster lookup | Slower |

---

## 10. Common Mistakes (Exam-Important)

1. Using indexing on set ❌
2. Expecting order ❌
3. Using `{}` for empty set ❌
4. Removing non-existing element with `remove()` ❌

---

## 11. One-line Exam Definition

> **A set is an unordered, mutable collection of unique elements in Python.**

---

## 12. Must-Remember Points

* Sets remove duplicates automatically
* No indexing or slicing
* Use for membership testing and set operations
* `{}` ≠ empty set


In [ ]:
#Set. : {},unordered and immutable, but Add/Remove ok . No dublicatesk
fruits = {"apple","orange","banana","coconut","pineapple","dragan","kiwi"}
type(fruits)
dir(fruits)
help(fruits)# we can use lenght,banana in fruits.try once


Help on set object:

class set(object)
 |  set() -> new empty set object
 |  set(iterable) -> new set object
 |
 |  Build an unordered collection of unique elements.
 |
 |  Methods defined here:
 |
 |  __and__(self, value, /)
 |      Return self&value.
 |
 |  __contains__(...)
 |      x.__contains__(y) <==> y in x.
 |
 |  __eq__(self, value, /)
 |      Return self==value.
 |
 |  __ge__(self, value, /)
 |      Return self>=value.
 |
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |
 |  __gt__(self, value, /)
 |      Return self>value.
 |
 |  __iand__(self, value, /)
 |      Return self&=value.
 |
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |
 |  __ior__(self, value, /)
 |      Return self|=value.
 |
 |  __isub__(self, value, /)
 |      Return self-=value.
 |
 |  __iter__(self, /)
 |      Implement iter(self).
 |
 |  __ixor__(self, value, /)
 |      Return self^=value.
 |
 |  __le__(self, value, /

Below is a **clear, structured, and exam-oriented explanation** of **Tuples in Python**, with **simple examples**, **real-time use cases**, and **important exam points**.

---

## 1. What is a Tuple?

A **tuple** is an **ordered**, **immutable** collection used to store **multiple values** in a single variable.

```python
t = (10, 20, 30)
```

---

## 2. Why do we use Tuples?

We use tuples when:

* Data **should not change**
* We want **data safety**
* We need **faster access than lists**
* Representing **fixed records** (coordinates, dates, config values)

In short:

> **Use tuples for fixed, read-only data.**

---

## 3. Key Properties of Tuples (Very Important)

| Property          | Description          |
| ----------------- | -------------------- |
| Ordered           | Maintains position   |
| Immutable         | Cannot modify values |
| Allows duplicates | Same value allowed   |
| Indexed           | Supports indexing    |
| Faster than lists | Less memory          |

---

## 4. Creating Tuples

```python
t1 = (1, 2, 3)
t2 = ("Sundar", 25, True)
```

### Single-element tuple (Important)

```python
t = (5,)   # comma is mandatory
```

---

## 5. Accessing Tuple Elements

```python
t = (10, 20, 30, 40)

print(t[0])    # 10
print(t[-1])   # 40
```

---

## 6. Tuple Slicing

```python
print(t[1:3])
```

Output:

```
(20, 30)
```

---

## 7. Immutability (Key Concept)

❌ Not allowed:

```python
t[1] = 50   # Error
```

✔ Allowed:

```python
t = t + (50,)   # creates new tuple
```

---

## 8. Tuple Methods (Very Few)

| Method     | Use               |
| ---------- | ----------------- |
| `count(x)` | Count occurrences |
| `index(x)` | Find index        |

Example:

```python
t = (1, 2, 2, 3)
print(t.count(2))   # 2
```

---

## 9. Looping Through a Tuple

```python
for x in t:
    print(x)
```

---

## 10. Real-Time Examples

### 1. Coordinates

```python
point = (10, 20)
```

---

### 2. Database record

```python
student = (101, "Sundar", 85)
```

---

### 3. Function returning multiple values

```python
def calc(a, b):
    return a+b, a-b

result = calc(10, 5)
```

---

## 11. Tuple vs List (Exam-Important)

| Feature  | Tuple      | List         |
| -------- | ---------- | ------------ |
| Mutable  | ❌ No       | ✔ Yes        |
| Syntax   | `()`       | `[]`         |
| Methods  | Few        | Many         |
| Speed    | Faster     | Slower       |
| Use case | Fixed data | Dynamic data |

---

## 12. Common Mistakes

1. Forgetting comma in single-element tuple ❌
2. Trying to modify tuple ❌
3. Using tuple when data must change ❌

---

## 13. One-Line Exam Definition

> **A tuple is an ordered, immutable collection used to store multiple values in Python.**

---

### Easy rule to remember:

> **Tuple = Fixed data + No change**

In [ ]:
fruits = ("apple","orange","banana","coconut","pineapple","dragan","kiwi")
t =()   # creates new tuple
t = t + (50,67)   # creates new tuple
print(t)

(50, 67)


In [ ]:
#Shoping cart program Exercise
foods = []
prices = []
total = 0
while True :
  food = input("enter a food to buy or say q to enough : ")
  if food == "q":
    break
  else:
    price = int(input(f"enter tuhe price of {food}:rs"))
    foods.append(food)
    prices.append(price)
print("your cart")
for food in foods:
    print(food,end="-")
for price in prices:
  total += int(price)
print(f"your total {total} rs")

enter a food to buy or say q to enough : jerfh
enter tuhe price of jerfh:rs32
enter a food to buy or say q to enough : rfgrt
enter tuhe price of rfgrt:rs312
enter a food to buy or say q to enough : q
your cart
jerfh-rfgrt-your total 344 rs


In [ ]:
# 2d List
fruits = ["apple","orange"]
veggies = ["cucumber","potato"]
drinks = ["old mounk","kingfisher"]
hi = [fruits,veggies,drinks]
#print(hi[0])# try it to clerity
print(hi[0][1])

num_pad = ((1,2,3),(4,5,6),(7,8,9),["*",0,"#"])
for row in num_pad:
  for hi in row:
    print(hi,end = " ")
  print()

orange
1 2 3 
4 5 6 
7 8 9 
* 0 # 


In [ ]:
#pyhton quiz
questions = ("How are you doing ? : ","what are the life source ? :","what to do ? :")
options = (("A.fine","B.ok","C.bad"),("A.water","B.air","C.food"),("________________"))
answer = ("B","A","NOTHING")
guesses = []
score = 0
question_num = 0
for question in questions:
    print("======")
    print(question)
    for option in options [question_num]:
       print(option)
    guess = input("enter (A,B,C,D) : ").upper()
    guesses.append(guess)
    if guess == answer[question_num]:
      score += 1
      print("correct")
    else:
      print("wrong")
      print(f"{answer[question_num]} is the correct answer")
    question_num+= 1




How are you doing ? : 
A.fine
B.ok
C.bad
enter (A,B,C,D) : a
wrong
B is the correct answer
what are the life source ? :
A.water
B.air
C.food
enter (A,B,C,D) : b
wrong
A is the correct answer
what to do ? :
_
_
_
_
_
_
_
_
_
_
_
_
_
_
_
_
enter (A,B,C,D) : nothing
correct


Below is a **focused, intermediate-level explanation of `dictionaries` in Python**, aligned with **real-time usage, DSA relevance, data analytics, interview expectations, and common pitfalls**. This avoids beginner repetition and concentrates on **how dictionaries are actually used**.

---

## 1. What a Dictionary Really Is (Intermediate View)

A **dictionary** is a **hash-based mapping** of **keys → values** designed for:

* Fast lookups (average **O(1)**)
* Structured, labeled data
* Flexible, readable code

```python
employee = {
    "id": 101,
    "name": "Sundar",
    "role": "Analyst",
    "salary": 45000
}
```

---

## 2. Why Dictionaries Are Critical in Real Applications

Dictionaries are used when:

* Data has **meaningful labels**
* You need **fast access**
* Records may grow dynamically
* Data structure resembles **JSON / DB rows**

Almost every backend, analytics, and ML pipeline uses dictionaries.

---

## 3. Keys: The Most Important Concept

### Valid keys

* Immutable
* Unique
* Hashable

```python
"name"          # string
101             # integer
(2025, "Q1")    # tuple
```

### Invalid keys

```python
[1, 2]      # list
{1, 2}      # set
```

**Interview rule:**

> If it can change, it cannot be a dictionary key.

---

## 4. Real-Time Dictionary Examples

### 1. Database-Like Record

```python
user = {
    "user_id": 5001,
    "email": "abc@gmail.com",
    "active": True
}
```

Used in:

* Backend APIs
* ORM results
* JSON responses

---

### 2. Frequency Counting (Very Common)

```python
data = ["a", "b", "a", "c", "b", "a"]
freq = {}

for item in data:
    freq[item] = freq.get(item, 0) + 1
```

Used in:

* Text analytics
* Log processing
* Interview problems

---

### 3. Configuration Management

```python
config = {
    "host": "localhost",
    "port": 3306,
    "debug": False
}
```

Why dictionary?

* Named values
* Easy access
* Readable configuration

---

## 5. Dictionary Methods (Intermediate Use)

```python
d.keys()      # all keys
d.values()    # all values
d.items()     # key-value pairs
d.get(k, v)   # safe access
d.pop(k)      # remove key
d.update()    # merge dictionaries
```

### Safe access (important)

```python
age = d.get("age", 0)
```

Avoids `KeyError`.

---

## 6. Iterating Dictionaries Correctly

```python
for key, value in d.items():
    print(key, value)
```

Better than:

```python
for key in d:
    print(key, d[key])
```

---

## 7. Dictionaries in DSA (Core Use)

### Graph Representation

```python
graph = {
    1: [2, 3],
    2: [4],
    3: []
}
```

### Memoization (Dynamic Programming)

```python
memo = {}
```

Used to:

* Avoid recomputation
* Improve time complexity

---

## 8. Dictionary Comprehension (Intermediate Skill)

```python
squares = {x: x*x for x in range(1, 6)}
```

Used in:

* Data transformation
* Analytics pipelines
* Cleaner code

---

## 9. Nested Dictionaries (Real-World Data)

```python
employee = {
    "id": 101,
    "personal": {
        "name": "Sundar",
        "city": "Vizag"
    },
    "skills": ["Python", "SQL"]
}
```

This structure mirrors:

* JSON
* API payloads
* NoSQL documents

---

## 10. Dictionaries vs Lists (Decision Rule)

| Use Case           | Choose     |
| ------------------ | ---------- |
| Ordered items      | List       |
| Labeled data       | Dictionary |
| Fast lookup        | Dictionary |
| Index-based access | List       |

---

## 11. Performance Insight (Interview Level)

* Dictionary lookup → **O(1)** average
* List search → **O(n)**

That is why:

* Dictionaries are preferred for lookups
* Sets/dicts are used in DSA problems

---

## 12. Common Intermediate Mistakes

❌ Accessing missing key directly

```python
d["age"]   # KeyError
```

✅ Use `get()`

❌ Using mutable keys
❌ Confusing `keys()` with values
❌ Overwriting keys unintentionally

---

## 13. When Interviewer Expects Dictionary Usage

* Counting problems
* Mapping relationships
* Graphs
* Caching / memoization
* JSON-like data handling

---

## 14. One-Line Intermediate Summary

> **Dictionary = fast lookup + labeled data + real-world structure**

---

If you want next:

* **Advanced dictionary interview questions**
* **DSA problems using dictionaries**
* **Set vs dictionary tricky questions**
* **Analytics case study using dictionaries**

Tell me what to continue.


Understood. Below is the **same dictionary content**, but now I will **slow down and explain the SYNTAX in detail**, line-by-line, **why each part exists**, **how Python reads it**, and **how you should think about it**. This is **intermediate-level syntax understanding**, not just definitions.

---

# DICTIONARIES — SYNTAX EXPLAINED IN DEPTH

---

## 1. Dictionary Creation Syntax

### Basic syntax

```python
employee = {
    "id": 101,
    "name": "Sundar",
    "role": "Analyst",
    "salary": 45000
}
```

### Syntax Breakdown

| Part       | Meaning                                          |
| ---------- | ------------------------------------------------ |
| `employee` | Variable name (reference to dictionary object)   |
| `=`        | Assignment operator                              |
| `{ }`      | Curly braces → tells Python this is a dictionary |
| `"id"`     | Key (must be immutable)                          |
| `:`        | Separator between key and value                  |
| `101`      | Value                                            |
| `,`        | Separates key–value pairs                        |

**How Python reads this internally:**

> “Create a hash table where each key maps to a value.”

---

## 2. Accessing Values (Index vs Key Concept)

### Correct way

```python
employee["name"]
```

### Why this syntax?

* Dictionaries do **NOT** use index numbers
* Values are accessed using **keys**

| Code       | Meaning    |
| ---------- | ---------- |
| `employee` | Dictionary |
| `["name"]` | Key lookup |

Python:

> “Find the hash of key `'name'` and return its value.”

---

## 3. Using `get()` — Safe Access Syntax

```python
employee.get("age", 0)
```

### Syntax Explained

| Part    | Meaning                        |
| ------- | ------------------------------ |
| `get()` | Dictionary method              |
| `"age"` | Key to search                  |
| `0`     | Default value if key not found |

### Why use `get()`?

* Prevents `KeyError`
* Cleaner logic
* Used in production code

---

## 4. Adding a New Key–Value Pair

```python
employee["department"] = "Data"
```

### Syntax Meaning

| Part                     | Meaning    |
| ------------------------ | ---------- |
| `employee["department"]` | New key    |
| `=`                      | Assignment |
| `"Data"`                 | Value      |

Python:

> “If key exists → update, else → create new key.”

---

## 5. Updating an Existing Key

```python
employee["salary"] = 50000
```

Same syntax as adding — Python decides automatically.

---

## 6. Removing Elements — Syntax Differences

### `pop()` (Preferred)

```python
employee.pop("salary")
```

| Part       | Meaning       |
| ---------- | ------------- |
| `pop()`    | Removes key   |
| `"salary"` | Key to remove |

Returns the removed value.

---

### `del`

```python
del employee["role"]
```

* Deletes key
* No return value

---

## 7. Iterating a Dictionary (Very Important)

### Using `items()`

```python
for key, value in employee.items():
    print(key, value)
```

### Syntax Explained

| Part         | Meaning                      |
| ------------ | ---------------------------- |
| `items()`    | Returns `(key, value)` pairs |
| `key, value` | Tuple unpacking              |
| `for`        | Loop over pairs              |

This is the **most correct way**.

---

## 8. Dictionary Comprehension — Syntax Deep Dive

```python
squares = {x: x*x for x in range(1, 6)}
```

### Breakdown

| Part                  | Meaning                  |
| --------------------- | ------------------------ |
| `{}`                  | Dictionary comprehension |
| `x`                   | Key                      |
| `:`                   | Key–value separator      |
| `x*x`                 | Value                    |
| `for x in range(1,6)` | Loop                     |

Python:

> “For each x, create key x and value x*x.”

---

## 9. Nested Dictionary Syntax

```python
employee = {
    "id": 101,
    "personal": {
        "name": "Sundar",
        "city": "Vizag"
    }
}
```

### Accessing nested data

```python
employee["personal"]["city"]
```

**Read it as:**

> employee → personal → city

---

## 10. Dictionary in DSA (Syntax View)

### Graph Representation

```python
graph = {
    1: [2, 3],
    2: [4],
    3: []
}
```

| Part    | Meaning        |
| ------- | -------------- |
| `1`     | Node           |
| `[2,3]` | Adjacent nodes |
| `:`     | Mapping        |

---

## 11. Frequency Counter — Syntax Logic

```python
freq = {}

for ch in data:
    freq[ch] = freq.get(ch, 0) + 1
```

### Step-by-step Execution

1. `freq.get(ch, 0)` → fetch count or 0
2. `+ 1` → increment
3. Assign back to `freq[ch]`

---

## 12. Common Syntax Errors (Must Know)

❌ Forgetting quotes

```python
employee[name]   # ERROR
```

❌ Using mutable key

```python
{[1,2]: "data"}  # ERROR
```

❌ Expecting index access

```python
employee[0]      # ERROR
```

---

## 13. Mental Model (Very Important)

Think of dictionary as:

> **Label → Value**

NOT:

> Position → Value

---

## 14. One-Line Exam Explanation

> A dictionary in Python is a hash-based collection that maps immutable keys to values, enabling fast lookup, insertion, and deletion.

---

If you want, next I can:

* Explain **every dictionary method syntax one-by-one**
* Solve **DSA problems using dict (with syntax reasoning)**
* Explain **sets syntax the same way**
* Give **exam-ready answers + code**

Just tell me what to continue.

```



In [ ]:
capitals = {"India" : "New Delhi","Russia" : "Moscow","Isarel" : "Gaza","japan" :"Tokyo"}
#print(dir(capitals))
#print(capitals.get("India",2))
#if capitals.get("India"):
  #print(f"welcome buddy")
#else :
 # print("guest")
#capitals.update({"Germany" : "Berlin"})
#capitals.update({"Russia" : "Berlin"})
#capitals.pop("japan")
#capitals.popitem()  # Clear try tis function
#print(capitals)
#hr = capitals.keys()
#print(hr)

#for key in capitals.keys():
 # print(key)
#values = capitals.values()
#print(values)
#for value in capitals.values():
 # print(value)
items = capitals.items()
print(items)



data = ["a", "b", "a", "c", "b", "a"]
freq = {}

for item in data:
    freq[item] = freq.get(item, 0) + 1
    print(type(item))

dict_items([('India', 'New Delhi'), ('Russia', 'Moscow'), ('Isarel', 'Gaza'), ('japan', 'Tokyo')])
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


In [ ]:
#concession stand program
menu = {"samosa" : 3.45,
        "jilebi" : 23.32,
        "chips" : 23.2,
        "soad" : 23.3,
        "nachos" : 3.3}
print("---------menu----------")
cart = []
total = 0
for key , value in menu.items():
  print(f"{key:10} : Rs {value}")
print("------------------------")
while True:
  food = input("select the items in menu and say ok to quit : ")
  if food == "ok":
    break
  elif menu.get(food) is not None:
    cart.append(food)
print(cart)
for food in cart:
 total += menu.get(food)
 print(food,end="-")
print()
print(f"Total is : Rs{total}")

---------menu----------
samosa     : Rs 3.45
jilebi     : Rs 23.32
chips      : Rs 23.2
soad       : Rs 23.3
nachos     : Rs 3.3
------------------------
select the items in menu and say ok to quit : samosa
select the items in menu and say ok to quit : chips
select the items in menu and say ok to quit : soad
select the items in menu and say ok to quit : ok
['samosa', 'chips', 'soad']
samosa-chips-soad-
Total is : Rs49.95


Here is a **clean, clear, and interview-ready table** that explains **when to use `list`, `tuple`, `set`, and `dictionary`**. You can **remember this easily** and **write it in exams**.

---

## When to Use Python Data Structures (Table Format)

| Feature / Need        | List   | Tuple  | Set           | Dictionary   |
| --------------------- | ------ | ------ | ------------- | ------------ |
| Data stored as        | Values | Values | Unique values | Key → Value  |
| Order matters         | ✅ Yes  | ✅ Yes  | ❌ No          | ✅ Yes (3.7+) |
| Allows duplicates     | ✅ Yes  | ✅ Yes  | ❌ No          | ❌ Keys: No   |
| Mutable (changeable)  | ✅ Yes  | ❌ No   | ✅ Yes         | ✅ Yes        |
| Indexed access        | ✅ Yes  | ✅ Yes  | ❌ No          | ❌ No         |
| Fast lookup           | ❌ No   | ❌ No   | ✅ Yes         | ✅ Yes        |
| Uses keys             | ❌ No   | ❌ No   | ❌ No          | ✅ Yes        |
| Supports mapping      | ❌ No   | ❌ No   | ❌ No          | ✅ Yes        |
| Can be dictionary key | ❌ No   | ✅ Yes  | ❌ No          | ❌ No         |
| Memory efficient      | ❌      | ✅      | ⚠ Medium      | ⚠ Medium     |

---

## When to Use (Decision Table)

| Situation                | Best Choice |
| ------------------------ | ----------- |
| Ordered, changeable data | List        |
| Fixed, ordered data      | Tuple       |
| Unique elements only     | Set         |
| Key–value relationship   | Dictionary  |
| Fast membership checking | Set         |
| Fast value lookup        | Dictionary  |
| Database-like record     | Dictionary  |
| Coordinates / constants  | Tuple       |

---

## Real-World Examples Table

| Use Case            | Data Structure | Example                       |
| ------------------- | -------------- | ----------------------------- |
| Shopping cart       | List           | `["milk", "bread"]`           |
| GPS location        | Tuple          | `(17.38, 78.48)`              |
| Unique user IDs     | Set            | `{101, 102, 103}`             |
| Employee record     | Dictionary     | `{"id":101, "name":"Sundar"}` |
| Frequency count     | Dictionary     | `freq[word] += 1`             |
| Visited nodes (DSA) | Set            | `visited = set()`             |

---

## Interview One-Line Summary Table

| Data Structure | One-Line Purpose               |
| -------------- | ------------------------------ |
| List           | Ordered, changeable collection |
| Tuple          | Ordered, fixed collection      |
| Set            | Unordered, unique elements     |
| Dictionary     | Fast key–value mapping         |

---

### Tip for Exams & Interviews

If **order matters → list/tuple**
If **uniqueness matters → set**
If **meaning/labels matter → dictionary**

---

If you want, next I can:

* Convert this into **MCQs**
* Create a **flowchart**
* Give **DSA problems per structure**
* Provide **memory & time complexity



In [ ]:
import random
import string
chars = string.punctuation + string.digits + string.ascii_letters
print(chars)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ


Below is a **clear, structured, and complete explanation of functions in Python**, from **basic idea to intermediate usage**, with **syntax explained step by step**, **real-time examples**, and **common mistakes**—the same depth you preferred earlier.

---

# FUNCTIONS IN PYTHON

---

## 1. What Is a Function?

A **function** is a **reusable block of code** that performs a **specific task**.

Instead of writing the same code again and again, you:

* Write it **once**
* Call it **whenever needed**

---

## 2. Why Functions Are Important

Functions help to:

* Reduce code repetition
* Improve readability
* Make debugging easier
* Organize large programs
* Reuse logic (very important in projects & interviews)

---

## 3. Basic Function Syntax

```python
def function_name(parameters):
    statements
    return value
```

### Syntax Explained

| Part            | Meaning                      |
| --------------- | ---------------------------- |
| `def`           | Keyword to define a function |
| `function_name` | Name of the function         |
| `(parameters)`  | Inputs to the function       |
| `:`             | Starts function body         |
| Indentation     | Function block               |
| `return`        | Sends result back            |

---

## 4. Simple Example

```python
def greet():
    print("Hello, Sundar")
```

### Calling the function

```python
greet()
```

### Output

```
Hello, Sundar
```

---

## 5. Function with Parameters

```python
def greet(name):
    print("Hello", name)
```

### Call

```python
greet("Sundar")
```

### Explanation

* `name` is a parameter
* `"Sundar"` is an argument

---

## 6. Function with Return Value

```python
def add(a, b):
    return a + b
```

### Call

```python
result = add(10, 20)
print(result)
```

### Why `return`?

* Sends value back to caller
* Stops function execution

---

## 7. Difference: `print()` vs `return`

| print()            | return           |
| ------------------ | ---------------- |
| Displays output    | Sends value back |
| For user           | For program      |
| Cannot reuse value | Can reuse        |

```python
def wrong():
    print(10)

def correct():
    return 10
```

---

## 8. Function with Default Arguments

```python
def power(base, exp=2):
    return base ** exp
```

### Calls

```python
power(5)      # 25
power(5, 3)   # 125
```

---

## 9. Keyword Arguments

```python
def student(name, age):
    print(name, age)

student(age=23, name="Sundar")
```

Order does not matter.

---

## 10. Variable Scope (Very Important)

```python
x = 10

def demo():
    x = 5
    print(x)

demo()
print(x)
```

### Output

```
5
10
```

* Inside function → local variable
* Outside → global variable

---

## 11. `*args` (Multiple Positional Arguments)

```python
def total(*numbers):
    return sum(numbers)

total(10, 20, 30)
```

* `*args` collects arguments into a tuple

---

## 12. `**kwargs` (Multiple Keyword Arguments)

```python
def details(**info):
    print(info)

details(name="Sundar", age=23)
```

* `**kwargs` collects into dictionary

---

## 13. Real-Time Use of Functions

### Example: Validation Logic

```python
def is_adult(age):
    return age >= 18
```

### Example: Data Processing

```python
def clean_data(data):
    return list(set(data))
```

### Example: DSA

```python
def binary_search(arr, target):
    pass
```

---

## 14. Built-in vs User-Defined Functions

| Type         | Example            |
| ------------ | ------------------ |
| Built-in     | `print()`, `len()` |
| User-defined | `add()`, `greet()` |

---

## 15. Common Function Mistakes

❌ Forgetting to call function
❌ Using `print` instead of `return`
❌ Wrong indentation
❌ Returning nothing unintentionally (`None`)

```python
def demo():
    a = 10   # no return → returns None
```

---

## 16. Interview One-Line Definition

> A function is a reusable block of code that performs a specific task and optionally returns a result.

---

## 17. When to Use Functions (Decision)

| Situation      | Use Function |
| -------------- | ------------ |
| Repeated logic | ✅            |
| Large program  | ✅            |
| Testing        | ✅            |
| Readability    | ✅            |

---

If you want next, I can:

* Explain **lambda functions**
* Explain **recursive functions**
* Explain **return vs yield**
* Solve **function-based interview problems**

Just tell me what to continue.


In [ ]:
#Function  = A block of reuseable code
# Place () After the function name to invoke it
#def engage(name):
 # print(f"hi how are {name}")
  #print("when will you come")
  #print("soon after the pongal")
#engage("jack")
print()
#engage("sundar")
#def engage(name,date):# Arugument
 # print(f"hi how are {name}")
  #print(f"i will come on {date}")
  #print("soon after the pongal")
#engage("jack","434-432-234")
#print()
#engage("sundar","32-343-344")

#def display_invoice(username,amount,due_date):
 # print(f"Hello{username}")

  #print(f"your bill of Rs {amount:.3f} and lastdate {due_date}")
#display_invoice("Sundar",32.2434,"32-42-2424")

#Return = statement used to end a function and send a result back to the caller
#def add(x , y):
 # g = x + y
 # return g
#def subtract (x , y):
 # g = x - y
  #return g
#def multiple (x , y):
  #g = x * y
  #return g
#def divide (x , y):
  #g = x // y
  #return g
#print(add(1,2))
#print(subtract(1,2))
def create_name(first,last):
  first = first.capitalize()
  last = last.capitalize()
  return first + " " +last
full_name = create_name("Soma","Sundar")
print(full_name)



Soma Sundar


In [ ]:
# Default arguments = A default value for certain parameters,,default is used when that argument is omitted make your functions more flexible,reduces of arguments
#  1)Positional 2)Default 3)keyboard 4)arbitrary

#def net_price(list_price , discount ,tax):
 # return list_price * (1 - discount) * (1 + tax)
#net_price(89,0,3)

#def net_price(list_price , discount = 2 ,tax = 0.9):
 # return list_price * (1 - discount) * (1 + tax)
#net_price(89)

# Exercise :
import time
def count(start,end): #  we cant give default valur to the frist argument
  for x in range (start,end+1):
    print(x)
    time.sleep(1)
  print("wake up")
count(3,4)
#Why?
#The time module doesn't have a function called nap()
#sleep() is the actual function name.


3
4
wake up


In [ ]:
# Key word arguments = An Argument preceded by an identifier,,help with readability order of the arguments doesnt matter
#def hello(greeting,title,first,last):
 # print(f"{greeting} {title} {first} {last}")
#hello(greeting="rfe",first="dfer",title="refer",last="erfer")
def get_phone():
  input("G")


1 2 3 4 5 6 7 8 9 10 

In [ ]:
#Args  = allows you to pass multiple non-key arguments
# key arguments = allows you to pass multiple keyword argument
#def add(*args):
 # total = 0
  #for arg in args:
   # total += 100
  #return total
#print((1,2,3))

#def add(*nums):
 # total = 0
  #for num in nums:
   # total += 2
  #return total
#print(add(3))


#def display_name(*args):
 # for arg in args:
  #  print(arg,end = " ")
#display_name("hi dude","")

#keyword argument

def print_address(**kwargs):
  print(type(kwargs))
print_address(street = "chinna veedhi",city = "vzm",state = "AP",zip = "21313")

#  Exercise
def shipping_label(*args,**kwargs):
  for arg in args:
    print(arg, end = " ")
  print()
  print(f"{kwargs.get('street')}")
  for value in kwargs.values():
    print(value , end = " ")

shipping_label("DR","Sundar","Soma","eeeccc",
               street = "chinna",phone = "00000000",city = "vzm")

<class 'dict'>
DR Sundar Soma eeeccc 
chinna
chinna 00000000 vzm 

In [ ]:
# iterable = An object / collection that can return its elements one at a time
# , allowing it to be iterated over  in a loop
#numbers = [1,2,3,4,5,5]
#for number in numbers:
 # print(number)
  # Sets shouldnt be reverseable
my = {"A" : 1,"B" : 2,"C": 3}
for key , value in my.items():# only keys key function,,,only values = value fuction,,,both = items()
  print(key,value)


A 1
B 2
C 3


In [ ]:
# membership operators = used to test wheather a value or a variable is found in a sequence in string,list,tuple,sets,dct
#word = "Apple"
#hi = input ("enter the letter in a secret word : ")
#if hi not in word : # observe in
 # print(f"there is a {hi}")
#else:
 # print("no")

#students = {"soma","sundar","mahi","jaddu"}
#student = input("Enter Name : ")
#if student not in students :
 # print(f"{student} welcome")
#else :
  #print("evadu bey nuvu salle")

#grades = {"soma" : "AA","SUNDAR" : "BBB","FUNCTION" : "YYYYY"}
#student = input("enter the student name : ")
#if student in grades:
 # print(f"pakka poyyi aduko {student} grade {grades[student]}")
#else :
 # print("dobay")

email = "sundarkalla987@gmail.com"
if "@" in email and "." in email:
  print("ok")
else :
  print("fuck")


ok


In [ ]:
#List comprehension  = A concise wat to create lists in python
# Compact and easier to read than trditional loops [Expression for value in iterable if conditon]
#doubles = []
#for x in range(1,34):
 # doubles.append(x*2)
#print(doubles)

#doubles = [x * 2 for x in range(1,11)] #This is called list comphrensive
#print(doubles)

#doubles = [x * 2 for x in range(1,11)]
#triples = [y*3 for y in range(1,4)]
#print(triples)

# Strings :
#fruits = ["appple","mango","orange","coconut"]
#fruits = [fruit.upper() for fruit in fruits]
#print(fruits)

#fruits = ["appple","mango","orange","coconut"]
#fruits = [fruit[0] for fruit in fruits]
#print(fruits)

# conditions
numbers = [-1,2,-3,4,-5]
pos_num = [num for num in numbers if num >= 0]
even_num = [num for num in numbers if num % 2 == 0]
print(even_num)


[2, 4]


In [ ]:
# match case statement = An Alternative to using many "elif " sttements execute some code if a value matches a "case"
# cleaner and syntax is more readable.
def day_of_week(day):
 match day:
  case 1:
   return "It is Sunday"
  case 2:
   return "It is Monday"
  case 3:
    return "It is Tuesday"
  case 4:
     return "It is Wednesday"
  case 5:
    return "It is Thursday"
  case 6:
    return "It is Friday"
  case 7:
     return "It is Saturday"
  case _:
     return "Not a valid day"

print(day_of_week("pizza"))


Not a valid day


In [ ]:
# Module = a file containing code you wnat to include in your program
# use "import" to include a module (built in or your own)
# useful to break up a large program reuseable separete file.

# print(help("math"))

# import math
#import math as m
#from math import pi
#print(pi)
#a,b,c,d = 1,2,3,4

#import example
#example.square("hi this is sundar")

import Sundar
Sundar.cube(324)

34012224

In [ ]:
%%writefile Sundar.py
pi = 3.141
def square(x):
  return x ** 2
def cube (x):
  return x ** 3
def cricumferance(radius):
  return 2 * pi * radius

Writing Sundar.py


In [ ]:
#Varible scope = Where a varible is visible and accesible
#scope resolution = local-->eclosed-->global-->built-in

#def fun1():
 # a = 1 # local number
  #print(a)

#def fun2():
 # b = 2
  #print(b)
def fun1():
  x = 1 # local number
  def fun2():
    print(x)
  fun2()
fun1()

1


Below is a **clear, structured, and exam/interview-ready explanation of *Variable Scope in Python***, explained **step by step**, with **simple examples**, **rules**, and **common mistakes**.

---

# VARIABLE SCOPE IN PYTHON

---

## 1. What is Variable Scope?

**Variable scope** defines **where a variable can be accessed or used** in a program.

In simple words:

> Scope decides **which part of the program can see a variable**.

---

## 2. Types of Variable Scope in Python

Python follows the **LEGB rule**:

| Scope | Meaning   |
| ----- | --------- |
| L     | Local     |
| E     | Enclosing |
| G     | Global    |
| B     | Built-in  |

---

## 3. Local Scope

### Definition

A variable created **inside a function**.

```python
def demo():
    x = 10   # local variable
    print(x)

demo()
```

❌ This will cause error:

```python
print(x)   # NameError
```

### Key points

* Exists only inside the function
* Destroyed after function execution

---

## 4. Global Scope

### Definition

A variable created **outside all functions**.

```python
x = 20   # global variable

def demo():
    print(x)

demo()
print(x)
```

### Output

```
20
20
```

---

## 5. Modifying Global Variable (Important)

❌ Wrong way

```python
x = 10
def demo():
    x = x + 5   # Error
```

✅ Correct way (`global`)

```python
x = 10
def demo():
    global x
    x = x + 5

demo()
print(x)
```

---

## 6. Enclosing Scope (Nested Functions)

```python
def outer():
    y = 5
    def inner():
        print(y)
    inner()

outer()
```

* `y` is **enclosing variable**
* Accessible inside `inner()`

---

## 7. Modifying Enclosing Variable (`nonlocal`)

```python
def outer():
    y = 5
    def inner():
        nonlocal y
        y = y + 1
    inner()
    print(y)

outer()
```

---

## 8. Built-in Scope

Built-in functions and names.

```python
print(len("Python"))
```

* `print`
* `len`
* `range`

❌ Avoid overwriting:

```python
len = 10   # BAD practice
```

---

## 9. LEGB Rule (Search Order)

When Python looks for a variable, it searches in this order:

1. Local
2. Enclosing
3. Global
4. Built-in

---

## 10. Example Demonstrating LEGB

```python
x = "global"

def outer():
    x = "enclosing"
    def inner():
        x = "local"
        print(x)
    inner()

outer()
```

### Output

```
local
```

---

## 11. Common Mistakes

❌ Confusing local and global
❌ Forgetting `global` keyword
❌ Overwriting built-in names
❌ Expecting variable outside function to change automatically

---

## 12. Real-Time Use Cases

| Scenario              | Scope     |
| --------------------- | --------- |
| Temporary calculation | Local     |
| Configuration value   | Global    |
| Nested logic          | Enclosing |
| System functions      | Built-in  |

---

## 13. Interview One-Line Answer

> Variable scope defines the visibility and lifetime of a variable within different parts of a Python program.

---

## 14. Quick Summary Table

| Scope     | Defined Where     | Accessible Where |
| --------- | ----------------- | ---------------- |
| Local     | Inside function   | Same function    |
| Enclosing | Outer function    | Inner function   |
| Global    | Outside functions | Everywhere       |
| Built-in  | Python itself     | Everywhere       |

---

If you want next, I can:

* Explain **scope with diagrams**
* Solve **scope-based interview questions**
* Explain **global vs nonlocal clearly**
* Show **real project examples**

Just tell me what to continue.


In [ ]:
# Python credit card validator program
#Remove any "-" or '  '
#Add all digit in the odd places from right to left
# Double evey Every second from right to left .
 # (if result is a two digit number, add the two digit number together to get a single digit.)
# Sum the totals of steps 2 & 3
#if sum is divisible by 10,the credit card # is vaild
sum_odd_digits = 0
sum_even_digits = 0
total = 0
#card_number = int(input("enter a credit card number :  ")) int does NOT have .replace()
card_number = input("enter a credit card number :  ")#The replace() method requires TWO arguments:
#string.replace(old, new)
#But you provided only one ("-").
card_number = card_number.replace("-","")# practize the code by keep
card_number = card_number[::-1]
print(card_number)
#Sum of all odd places
for x in card_number[::2]:
  sum_odd_digits += int(x)
#3
for x in card_number[1::2]:
  x = int(x) * 2
  if x >= 10:

    sum_even_digits += (1+(x % 10))
  else:
    sum_even_digits += x
# 4
total = sum_odd_digits + sum_even_digits
#5
if total % 10 == 0:
  print("og")
else :
  print("noonoono")


enter a credit card number :  378282246310005
500013642282873
og


In [3]:
def show_balance():
  print(f"Your Current balance is {balance:.2f}Rs")
def deposit():
  amount = float(input("Enter an amount to br deposited : "))
  if amount < 0:
    print("invalid deposit amount")
    return 0
  else:
    return amount
def withdraw():
  amount = float(input("Enter an amount that would be withdraw : "))
  if amount > balance:
    print("No MOney with that number")
    return 0
  elif amount < 0 :
    print("you cant withadraw")
    return 0
  else:
    return amount
balance = 0
is_running = True
while is_running:
  print("Banking System")
  print("Balance")
  print("Deposit")
  print("whithdraw")
  print("exit")

  choice = input("Enter your options (1-4)")
  if choice == "1":
    show_balance()
  elif choice == "2":
    balance += deposit()
  elif choice == "3":
   balance -= withdraw()
  elif choice == "4":
    is_running = False
  else:
    print("invalid")
print("Thank you for using")

Banking System
Balance
Deposit
whithdraw
exit
Enter your options (1-4)1
Your Current balance is 0.00Rs
Banking System
Balance
Deposit
whithdraw
exit
Enter your options (1-4)2
Enter an amount to br deposited : 1222
Banking System
Balance
Deposit
whithdraw
exit
Enter your options (1-4)3
Enter an amount that would be withdraw : 222
Banking System
Balance
Deposit
whithdraw
exit
Enter your options (1-4)1
Your Current balance is 1000.00Rs
Banking System
Balance
Deposit
whithdraw
exit
Enter your options (1-4)4
Thank you for using
